### PR TITLE
[CORE] Add new permission model

### PR DESCRIPTION
**Pull request by Memo2109**
@stefaus  @m273d15 ,

Thank you very much Stefan for the comments.
I clean the code. The code can now read much better than before.

### CODE REVIEW

**Comment by saros-infrastructure** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 38
           

Complexity increasing per file
==============================
- core/src/saros/session/UserPrivilege.java  3
- server/src/saros/server/console/ServerConsole.java  1
- core/test/junit/saros/session/SarosSessionUserTest.java  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/app/saros/saros/pullRequest?prid=4555642&bid=15321763)

File: core/src/saros/session/User.java [hunk 1]
```diff
@@ -54,7 +55,74 @@
...
+  @Deprecated
   private volatile Permission permission = Permission.WRITE_ACCESS;
+  
+  
+  
+  
+  
+  
+  /* More flexible Permissions also known as Privilege */
+  private Map<UserPrivilege.Keys, UserPrivilege> privileges;
```
Comment (srossbach) [inline core/src/saros/session/User.java]: Not thread safe.

File: core/src/saros/session/User.java [hunk 2]
```diff
@@ -54,7 +55,74 @@
...
+  public Map<UserPrivilege.Keys, UserPrivilege> getPrivileges() {
+      return this.privileges;
```
Comment (srossbach) [inline core/src/saros/session/User.java]: Map can be mutated by the caller.

File: core/src/saros/session/User.java [hunk 3]
```diff
@@ -54,7 +55,74 @@
...
+      this.privileges = privileges;
+  }
+  public void addPrivilege(UserPrivilege privilege) {
+      this.privileges.put(privilege.getKey(), privilege);
+  }
+  // get any privileges value or false
+  public boolean hasPrivilege(UserPrivilege.Keys privilege) {
+//System.out.println("3 - hasPrivilege() " + privilege + " : " + this.privileges.containsKey(privilege));
+	  if (this.privileges.containsKey(privilege)) {
+		  return this.privileges.get(privilege).getValue();
```
Comment (srossbach) [inline core/src/saros/session/User.java]: NPE

File: core/src/saros/session/User.java [hunk 4]
```diff
@@ -54,7 +55,74 @@
...
+	  }
+	  return false;
+  }
+
+  // convenience functions to privilege values
+  public boolean hasReadOnlyAccessPrivilege() {
```
Comment (srossbach) [inline core/src/saros/session/User.java]: I do not see the need for 1 line convenience methods. Using the enum Name should be sufficient.

File: core/src/saros/session/UserPrivilege.java [hunk 5]
```diff
@@ -0,0 +1,55 @@
+package saros.session;
+
+public class UserPrivilege {
```
Comment (srossbach) [inline core/src/saros/session/UserPrivilege.java]: Using an EnumMap in the User class is much easier than using nested Enums.

File: core/test/junit/saros/session/SarosSessionUserTest.java [hunk 6]
```diff
@@ -0,0 +1,133 @@
...
+  protected void setUp() {
+	  // System.out.println("1 - SarosSessionUserTest setUp()");
+  }
+  
+  // testing default privileges, all should return false
+  @Test
+  public void testSessionUserReadOnlyPrivilegeDefault() {
+	  setUp();
+// System.out.println("2 - testSessionUserReadOnlyPrivilegeDefault()" + HOST.hasReadOnlyAccessPrivilege());
+	  assertEquals(HOST.hasReadOnlyAccessPrivilege(), new Boolean(false));
```
Comment (srossbach) [inline core/test/junit/saros/session/SarosSessionUserTest.java]: assertTrue ?

File: core/test/junit/saros/session/SarosSessionUserTest.java [hunk 7]
```diff
@@ -0,0 +1,133 @@
...
+import org.junit.Before;
+
+import org.junit.Test;
+import saros.net.xmpp.JID;
+import saros.session.User;
+import saros.session.UserPrivilege;
+
+public class SarosSessionUserTest {
+
+  private JID ALICE = new JID("alice@test/Saros");
```
Comment (srossbach) [inline core/test/junit/saros/session/SarosSessionUserTest.java]: Naming convention violation.

File: core/test/junit/saros/session/SarosSessionUserTest.java [hunk 8]
```diff
@@ -0,0 +1,133 @@
...
+  @Test
+  public void testSessionUserDeleteSessionDataPrivilegeDefault() {
+	  assertEquals(HOST.hasDeleteSessionDataPrivilege(), new Boolean(false));
+  }
+  @Test
+  public void testSessionUserConfigureServerPrivilegeDefault() {
+	  assertEquals(HOST.hasConfigureServerPrivilege(), new Boolean(false));
+  }
+
+  // test add privilege
```
Comment (srossbach) [inline core/test/junit/saros/session/SarosSessionUserTest.java]: While you have add methods, you do not have any remove method but only a generic set method. So what about testing this behavior ?

File: server/src/saros/server/console/ServerConsole.java [hunk 9]
```diff
@@ -55,10 +56,11 @@ public void run() {
           List<String> cmd = Arrays.asList(line.split(" "));
           if (command.identifier().equals(cmd.get(0))) {
             command.execute(cmd.subList(1, cmd.size()), output);
+            found = true;
             break;
           }
         }
-        output.printf("Command '%s' is not known. (Type 'help' for available commands)", line);
+        if (found == false) output.printf("Command '%s' is not known. (Type 'help' for available commands)", line);
```
Comment (srossbach) [inline server/src/saros/server/console/ServerConsole.java]: ???

**Comment by saros-infrastructure** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 38
           

Complexity increasing per file
==============================
- server/src/saros/server/console/ServerConsole.java  1
- core/test/junit/saros/session/SarosSessionUserTest.java  1
- core/src/saros/session/UserPrivilege.java  3
         
```


See the complete overview on [Codacy](https://app.codacy.com/app/saros/saros/pullRequest?prid=4555642&bid=15321763)

**Comment by m273d15** [comment]
I just rebased the commits.

**Comment by saros-infrastructure** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 37
           

Complexity increasing per file
==============================
- core/test/junit/saros/session/SarosSessionUserTest.java  1
- core/src/saros/session/UserPrivilege.java  3
         
```


See the complete overview on [Codacy](https://app.codacy.com/app/saros/saros/pullRequest?prid=4555642&bid=15321763)

**Comment by saros-infrastructure** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 6
           

Complexity increasing per file
==============================
- server/src/saros/server/console/ServerConsole.java  1
- core/src/saros/session/UserPrivilege.java  1
- core/test/junit/saros/session/SarosSessionUserTest.java  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/app/saros/saros/pullRequest?prid=4555642&bid=15321763)

**Comment by saros-infrastructure** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 2
           

Complexity increasing per file
==============================
- server/src/saros/server/console/ServerConsole.java  1
- core/src/saros/session/UserPrivilege.java  1
- core/test/junit/saros/session/SarosSessionUserTest.java  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/app/saros/saros/pullRequest?prid=4555642&bid=15321763)

File: core/src/saros/session/User.java [hunk 10]
```diff
@@ -1,5 +1,7 @@
 package saros.session;
 
+import java.util.*;
```
Comment (tobous) [inline core/src/saros/session/User.java]: Our code should be formatted according to [google java style](https://google.github.io/styleguide/javaguide.html), which prohibits the use of [wildcard imports](https://google.github.io/styleguide/javaguide.html#s3.3.1-wildcard-imports).

In general, the code (including wildcard imports) should be formatted correctly by default when developing in Eclipse or IntelliJ environments that were set up according to our [guide](https://www.saros-project.org/contribute/development-environment.html). (Eclipse formats on saving, IntelliJ only formats when explicitly triggering it using the keyboard shortcut `ctrl+alt+l` (see [IntelliJ documentation](https://www.jetbrains.com/help/idea/reformat-and-rearrange-code.html)).)

**Comment by saros-infrastructure** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 3
           

Complexity increasing per file
==============================
- core/test/junit/saros/session/SarosSessionUserTest.java  1
- core/src/saros/session/UserPrivilege.java  1
- server/src/saros/server/console/ServerConsole.java  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/app/saros/saros/pullRequest?prid=4555642&bid=15321763)

**Comment by saros-infrastructure** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 2
           

Complexity increasing per file
==============================
- core/test/junit/saros/session/SarosSessionUserTest.java  1
- core/src/saros/session/UserPrivilege.java  1
- server/src/saros/server/console/ServerConsole.java  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/app/saros/saros/pullRequest?prid=4555642&bid=15321763)

**Comment by Memo2109** [comment]
Add UserPrivilege to session User. Enhanced code Quality and improved thread safety. Adds junit test for setting privilege

**Comment by m273d15** [comment]
We rebased the branch last week together. Now, the branch contains the same commits that are not related to your changes again. Please, clean your branch (rebase).

Please, solve the issues detected by codacy.
Our [guidelines](https://www.saros-project.org/contribute/guidelines.html) describe the commit tag schema. The tags describe the context (the component) of the changes within a commit. Please, don't introduce new tags as long as you don't introduce a complete new component.

**Comment by saros-infrastructure** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 4
           

Complexity increasing per file
==============================
- core/src/saros/session/UserPrivilege.java  1
- server/src/saros/server/console/ServerConsole.java  1
- core/test/junit/saros/session/SarosSessionUserTest.java  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/app/Saros/saros/pullRequest?prid=4555642&bid=15321763)

File: core/src/saros/session/User.java (hunk 2 above)
Comment (Memo2109) [reply core/src/saros/session/User.java]: I do not understand how a get method may mutate the instance? Do you think privileges will mutate?

**Comment by saros-infrastructure** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- core/test/junit/saros/session/SarosSessionUserTest.java  1
- core/src/saros/session/UserPrivilege.java  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/app/Saros/saros/pullRequest?prid=4555642&bid=15321763)

**Comment by saros-infrastructure** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 1
           

Complexity increasing per file
==============================
- server/src/saros/server/console/ServerConsole.java  1
- core/src/saros/session/UserPrivilege.java  1
- eclipse/src/saros/ui/views/SarosViewBrowserVersion.java  2
- core/test/junit/saros/session/SarosSessionUserTest.java  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/app/Saros/saros/pullRequest?prid=4555642&bid=15321763)

**Comment by saros-infrastructure** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 1
           

Complexity increasing per file
==============================
- core/src/saros/session/UserPrivilege.java  1
- core/test/junit/saros/session/SarosSessionUserTest.java  1
- eclipse/src/saros/ui/views/SarosViewBrowserVersion.java  2
- server/src/saros/server/console/ServerConsole.java  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/app/Saros/saros/pullRequest?prid=4555642&bid=15321763)

**Comment by m273d15** [comment]
What happened again with your git history 😿 ?

**Comment by saros-infrastructure** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 1
           

Complexity increasing per file
==============================
- server/src/saros/server/console/ServerConsole.java  1
- core/test/junit/saros/session/SarosSessionUserTest.java  1
- core/src/saros/session/UserPrivilege.java  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/app/Saros/saros/pullRequest?prid=4555642&bid=15321763)

**Pull request closed without merging**